tests/test_integration.py

```diff
@@ -772,6 +773,33 @@ def test_read_config_file_send_receive_cli(self):
         recv_messages = self.run_receiver(['--log-msgs', 'dict', '--conn-use-config-file'])
         self.assertTrue(len(sent_messages) == len(recv_messages) == 1)
 
+    def test_send_receive_on_release(self):
```

Comment:
I'm not sure what the test is testing here. It is testing just that the option might be used? If so, looks to me a bit complicated. How do you simulate the release of the message on the consumer site (to test the new code)? Could you please clarify it a bit?

Also, I don't understand the TOTAL_MSGS value, this is a fixed constant that should never be changed? ie.: If you use value of 200 instead of 100, then receiver's will be waiting for 400 messages but only 200 would be sent.

I believe that --action receiver's option might be useful for the test to trigger the new implemented path for released messages. It accepts the following values ('acknowledge', 'reject', 'release', 'noack').

Reply:
Ok, let me go through your questions.

1. In this test we have multiple receivers on a given address (balanced distribution), connected through Qpid Dispatch (which is listening at port 5673). At the moment a given receiver completes its work and closes its connection, the router might still be routing messages to that receiver and since connection / link was closed, the message is released. If that happens, cli-proton-python-sender gets stuck.
2. The TOTAL_MSGS is a const meant to represent the total amount of messages to be exchanged in the test. I can also add another variable to represent the number of receivers. But the reason is we need multiple receivers to reproduce the scenario, cause the issue eventually happens at disconnection from a receiver. So total message count divided by total receiver count will give expected amount of messages to be received by each instance. And without the '--on-release retry', if a message is released, receiver will get stuck and timeout.
3. I will evaluate the "--action release" and see if it helps reproducing the issue.

Reply:
@pematous @michalxo all requested changes have been done. I have also modified the test to force release of messages and ensure sender will be able to get all messages through the router to the connected receivers. Please let me know if there is anything else needed.